daemon/SecretsImpl/dataprotector.cpp

```diff
@@ -0,0 +1,326 @@
+        qWarning(lcSailfishSecretsDaemon) << "Could not decide which file is correct, there is no majority.";
+        return Irretrievable;
+    }
+
+    m_data = fileContents[maxOccourenceIndex];
+    *result = m_data;
+
+    return Success;
+}
+
+DataProtector::Status DataProtector::putData(const QByteArray &bytes)
+{
+    QDir dir(m_path);
+
+    // Check if the main directory exists, create if not
+    if (!dir.exists()) {
+        if (!dir.mkpath(m_path)) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't create protected root directory:" << m_path;
+            return ErrorCannotCreateDirectory;
+        }
+
+        qCDebug(lcSailfishSecretsDaemon) << "Protected root directory didn't exist, so the data is empty.";
+    }
+
+    QFileInfoList oldDirectories = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot);
+
+    QString dataDirName = QUuid::createUuid().toString().remove('{').remove('}');
+    QString dataDirPath = m_path + QStringLiteral("/") + dataDirName;
+
+    if (!dir.mkdir(dataDirPath)) {
+        qCWarning(lcSailfishSecretsDaemon) << "Can't create data directory:" << dataDirPath;
+        return ErrorCannotCreateDirectory;
+    }
+
+    QVector<QString> dataFilePaths = {
+        dataDirPath + QStringLiteral("/file0"),
+        dataDirPath + QStringLiteral("/file1"),
+        dataDirPath + QStringLiteral("/file2"),
+    };
+
+    QVector<QSharedPointer<QSaveFile> > files;
+    std::transform(dataFilePaths.begin(), dataFilePaths.end(), std::back_inserter(files), [](const QString &filePath) {
+        return QSharedPointer<QSaveFile>(new QSaveFile(filePath));
+    });
+
+    bool allOpen = std::all_of(files.begin(), files.end(), [](QSharedPointer<QSaveFile> &file) {
+        bool isOpen = file->open(QIODevice::WriteOnly);
+        if (!isOpen) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName();
+        }
+        return isOpen;
+    });
+
+    if (!allOpen) {
+        qCWarning(lcSailfishSecretsDaemon) << "Not all files could be open for writing.";
+        return ErrorCannotOpenFile;
+    }
+
+    for (QSharedPointer<QSaveFile> &file : files) {
+        qint64 bytesWritten = file->write(bytes);
+        if (bytesWritten != bytes.size()) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't write file:" << file->fileName();
+            return ErrorCannotWriteFile;
+        }
+        file->commit();
+    }
+
+    // Remove old data directories
+    for (QFileInfo &fileInfo : oldDirectories) {
+        QDir oldDataDir(fileInfo.absoluteFilePath());
+        oldDataDir.removeRecursively();
+    }
+
+    // Clear in-memory data, so it gets refreshed on next call
+    m_data.clear();
+    return Success;
+}
+
+
+DataProtector::Status DataProtector::processLegacy()
+{
+    int slashIndex = m_path.lastIndexOf('/');
+    if (slashIndex <= 0) {
+        return ErrorInvalidPath;
+    }
+
+    QString parentPath = m_path.mid(0, slashIndex);
```

Comment:
Shouldn't you use a ```QFileInfo(m_path).path()``` to do this? I understand that there is no chance to have '/' character in the name of the legacy file, but in case, and it's more syntaxically correct, I guess.

Reply:
Yep, good idea.